### ISSUE TITLE
Throttled field validation can happen after validation errors from submission have been added

### ISSUE THREAD

**Issue by dkoo761** [aldeed/meteor-autoform#668]
I have the following hook defined:

```javascript
AutoForm.hooks({
    createGroupForm: {
        onError: function (operation, result, template) {
            if (result.error === Constants.ERRORS.GROUP_ALREADY_EXISTS) {
                var errorKeys = [
                    { name: 'shortName', type: result.error}
                ];
                AutoForm.getValidationContext('createGroupForm').addInvalidKeys(errorKeys);
            }
        }
    }
});

var messages = {};
messages[Constants.ERRORS.GROUP_ALREADY_EXISTS] = 'A group with these initials already exists';

SimpleSchema.messages(messages);
```

When I submit the form by clicking the submit button and it triggers the GROUP_ALREADY_EXISTS error, I see the error message appear for a fraction of a second and then it disappears. If I click the form submit button a second time, the error appears again and this time it doesn't disappear which is the correct behaviour.

I asked myself: "what could be causing an invalid key to be removed the first time I submit the form?"

So I tried executing ```AutoForm.validateForm('createGroupForm')``` in my browser console to see what happens and of course this removes the invalid key from the the validation context (even though the field is technically still invalid based on server-side validation). This makes sense because AutoForm.validateForm() is only running validation on the client-side.

So my best guess is that somehow the form is being validated a second time immediately after the first validation. If I change the form to use validation="submit" instead of the default submitThenKeyup, then the problem disappears. I tried the other validation modes and the problem appears isolated to 4 validation modes: submitThenKeyup, submitThenBlur, keyup and blur.

**Comment by aldeed** [comment]
What form type? Reproduction code would be helpful so I can step through. I have very similar hooks that work in my own apps, but different form attributes can result in slightly different submission behavior. Most likely it thinks the form was successfully submitted so it's resetting it, but that shouldn't happen if `onError` is being called.

**Comment by dkoo761** [comment]
Here's the entire template if that helps. I have 2 forms with the same autoform attributes and the issue happens on both of them.

```html
<template name="createGroupModal">
    <div class="modal inmodal fade" id="modalCreateGroup" tabindex="-1" role="dialog"  aria-hidden="true" data-backdrop="static">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                    <h4 class="modal-title">Create Group</h4>
                </div>
                <div class="alert alert-danger hidden" role="alert"></div>
                {{#autoForm type="method" meteormethod="createGroup" schema=groupSchema id="createGroupForm" validation="submit" trimStrings=true}}
                <div class="modal-body">
                    {{> afQuickField name="name"}}
                    <div class="col-lg-2 col-m-2 col-sm-2 clear-side-padding">
                        {{> afQuickField name="shortName" label="Initials" placeholder="AAAA"}}
                    </div>
                    <div class="col-lg-offset-2 col-lg-8 col-m-offset-2 col-m-8 col-sm-offset-2 col-sm-8 clear-side-padding">
                        {{> afQuickField name="labelColor" noselect=true type="select-radio" template="buttonGroup" options=labelColorOptions value=labelColorDefault}}
                    </div>
                    {{> afQuickField name="orgId" type="hidden" label=false value=orgId}}
                </div>
                <div class="modal-footer">
                    <button type="button" class="btn btn-white" data-dismiss="modal">Cancel</button>
                    <button type="submit" class="btn btn-primary">Create</button>
                </div>
                {{/autoForm}}
            </div>
        </div>
    </div>
</template>
```
And the relevant schemas:

```javascript
Schema = {};

Schema.baseModel = new SimpleSchema({
    _id: {type: String, max: 50, regEx: SimpleSchema.RegEx.Id, optional: true},
    createdAt: {
        type: Date,
        autoValue: function () {
            if (!this.isSet) {
                return dateUtils.nowUtcAsDate();
            }
        }
    },
    updatedAt: {type: Date, optional: true},
    createdByUserId: {type: String, max: 50, regEx: SimpleSchema.RegEx.Id, optional: true},
    updatedByUserId: {type: String, max: 50, regEx: SimpleSchema.RegEx.Id, optional: true}
});

Schema.group = new SimpleSchema([Schema.baseModel, {
    orgId: {type: String, max: 50, regEx: SimpleSchema.RegEx.Id, optional: false},
    shortName: {type: String, max: 4, unique: true, optional: false},
    name: {type: String, max: 80, optional: false},
    labelColor: {type: String, max: 80, allowedValues: Constants.GROUP_LABEL_COLOR_CHOICES, optional: false}
}]);
```

**Comment by dkoo761** [comment]
Let me know if that's still not enough and I'll setup a minimal reproduction repo and send you the link.

**Comment by dkoo761** [comment]
I had some more time to look into this today. I narrowed it down so that the problem only occurs in this scenario:

1. User fills in the name field
2. User fills in the shortName field (which will cause server-side validation to fail)
3. User clicks the form submit button

When the user clicks the submit button, the 'submit form' event is triggered which calls _validateForm(), makes the method call to the server (which fails on duplicate shortName) and then I handle the error in the onError callback and the error message correctly appears in the form. The problem occurs because clicking the submit button also triggers a call to validateField (probably due to the blur event on the shortName field and/or maybe due to throttling of the validateField function) after the form has already been validated. This call to validateField then calls validateFormDoc() with 'key' set to 'shortName' which then wipes out the error message on the form because it validates the shortName field client-side against the schema (which passes validation as shortName will only fail server-side validation) and then erroneously removes shortName from _invalidKeys.

If I fill in the shortName field first and then the name field second and then click the form submit button, the same events are fired, but the error message is not wiped out after the call to validateField() because this time that call is only validating the name field and leaves the shortName field (and corresponding object in _invalidKeys) intact.

One possible solution might be to mark the form as 'submitted' at the end of the 'submit form' event handler and then have validateField check if the form is in the 'submitted' state and if so, prevent field validation from taking place since the form has already been validated. You could then clear the 'submitted' state from the form upon the next keyup or similar user event.

**Comment by aldeed** [comment]
@dkoo761, thank you for the good explanation. I explored some ways of mitigating, such as the workaround you mention, but I ended up deciding that this is simply a symptom of a larger problem, which is that there wasn't a way to attach sticky errors to a form, such that they don't get removed until you want them to.

So there's an experimental new feature in devel branch, where you can now call `this.addStickyValidationError` in a hook to do just that. This will be released with 5.0, probably later today or tomorrow.

**Comment by dkoo761** [comment]
Excellent, thanks.
